src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -124,10 +128,24 @@ bool PipelineExecutor::executeStep(std::atomic_bool * yield_flag)
     return false;
 }
 
+bool PipelineExecutor::checkTimeLimit()
+{
+    if (process_list_element)
+    {
+        bool continuing = process_list_element->checkTimeLimit();
+        // We call cancel here so that all processors are notified and tasks waken up
+        // so that the "break" is faster and doesn't wait for long events
+        if (!continuing)
+            cancel();
```

Comment:
well, I don't like we don't call cancel in case of exception from `process_list_element->checkTimeLimit()`.

What can we do:
* add softer check to `process_list_element`, something like `bool isTimeLimitExceeded()` which newer throws
* `PipelineExecutor::checkTimeLimit` uses this `isTimeLimitExceeded` version and just cancel. So, it's like a soft check.
* In other places like `PipelineExecutor::finalizeExecution` we call regular `process_list_element->checkTimeLimit()` and throw exception if needed.

Reply:
What do you think of having something like:

```c++
bool PipelineExecutor::checkTimeLimitSoft()
{
    try
    {
        return checkTimeLimit();
    }
    catch (...)
    {
        cancel();
        return false;
    }
}
```

Then calling `checkTimeLimitSoft` inside the `PullingPipelineExecutor` and `PullingAsyncPipelineExecutor` and in the `executeStepImpl` loop, and `checkTimeLimit` and the start and end of the execution. I think that fits what you were saying right?

Reply:
Yes, the idea is the same.
But let's try to avoid `try/catch(...)` with ignoring exception if possible :)